File: BEmu/MarketDataRequest/ElementMarketExceptionsArray.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="MarketDataRequest/ElementMarketExceptionsArray.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.MarketDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ElementMarketExceptionsArray : Element
    {
        private readonly List<ElementMarketExceptions> _exceptions;

        public ElementMarketExceptionsArray(IEnumerable<string> badFields)
        {
            this._exceptions = new List<ElementMarketExceptions>();
            foreach (var item in badFields)
            {
                this._exceptions.Add(new ElementMarketExceptions(item));
            }
        }

        public override Element GetValueAsElement(int index) { return (Element)this._exceptions[index]; }
        public override object this[int index] { get { return this._exceptions[index]; } }

        public override Name Name { get { return new Name("exceptions"); } }
        public override int NumElements { get { return 0; } }
        public override int NumValues { get { return this._exceptions.Count; } }
        public override bool IsComplexType { get { return false; } }
        public override bool IsArray { get { return true; } }
        public override bool IsNull { get { return false; } }
        public override IEnumerable<Element> Elements { get { if (this._exceptions.Count < -1) yield return null; } } //This is the only way I can think to have the compiler output "Enumeration yielded no results"

        public override object GetValue(int index)
        {
            return this._exceptions[index];
        }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}{1}[] = {{{2}", tabs, this.Name, Environment.NewLine);
            foreach (var item in this._exceptions)
            {
                result.Append(item.PrettyPrint(tabIndent + 1));
            }
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);

            return result;
        }


    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/MarketDataRequest/ElementMarketExceptionsArray.java
package com.bemu.BEmu.MarketDataRequest;

import java.util.List;
import java.util.ArrayList;
import com.bemu.BEmu.Element;
import com.bemu.BEmu.Name;

public class ElementMarketExceptionsArray extends ElementParent
{
    private final List<ElementMarketExceptions> _exceptions;

    public ElementMarketExceptionsArray(List<String> badFields)
    {
        this._exceptions = new ArrayList<ElementMarketExceptions>();
        
        for(int i = 0; i < badFields.size(); i++)
        {
        	String item = badFields.get(i);
        	this._exceptions.add(new ElementMarketExceptions(item));
        }
    }
	
	public Element getValueAsElement(int index)
	{
		return this._exceptions.get(index);
	}
	
    public Name name()
    {
    	return new Name("exceptions");
    }
    
    public int numValues()
    {
    	return this._exceptions.size();
    }
    
    public int numElements()
    {
    	return 0;
    }
    
    public boolean isComplexType()
    {
    	return false;
    }
    
    public boolean isArray()
    {
    	return true;
    }
    
    public boolean isNull()
    {
    	return false;
    }
	
    public Object getValue(int index)
    {
    	return this._exceptions.get(index);
    }
    
    protected StringBuilder prettyPrint(int tabIndent)
    {
        String tabs = com.bemu.BEmu.types.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

        result.append(String.format("%s%s[] = {%s", tabs, this.name().toString(), System.getProperty("line.separator")));
        
        for(int i = 0; i < this._exceptions.size(); i++)
        {
        	result.append(this._exceptions.get(i).prettyPrint(tabIndent + 1));
        }

        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));
        return result;
    }
	
	
}
